validator/client/src/main/java/tech/pegasys/teku/validator/client/ValidatorIndexProvider.java

```diff
@@ -106,4 +107,8 @@ private void logNewValidatorIndices(final Map<BLSPublicKey, Integer> knownValida
         __ ->
             publicKeys.stream().flatMap(key -> getValidatorIndex(key).stream()).collect(toList()));
   }
+
+  public SafeFuture<Map<BLSPublicKey, Integer>> getValidatorIndexesByPublicKey() {
+    return firstSuccessfulRequest.thenApply(__ -> unmodifiableMap(validatorIndexesByPublicKey));
+  }
```

Comment:
I'm not entirely convinced this is right - normally the idea is that you specify the public keys you want so the this provider knows what keys it should be requesting and updates them if needed. I'm very suspicious this is breaking the model somehow...

Reply:
I understand what you're saying. I don't have to rely on his internal state and I have to provide the collection of pubkeys. Even if wrt the current implementation things do not change that much (`getValidatorIndices` do not query for missing pubkeys at runtime),  it changes the assumptions of the provider.

Reply:
It may be sensible to refactor things so that `ValidatorIndexProvider` has a reference to `OwnedValidators` itself and then it wouldn't need to have the requested validators passed in which could simplify some things, but I think I'd tackle that as a separate PR as we'll have to be a bit careful about it.

Reply:
do you intend something like: https://github.com/ConsenSys/teku/pull/4917/commits/bf349240a4c4eec1d167348dc21a85b84987b423

Reply:
Yep that looks right. Surprised it didn't involve more changes actually...